File: hosts/android/feature/products/impl/src/main/java/io/paritytech/polkadotapp/feature_products_impl/domain/truapi/ProductPocketHostBridge.kt
package io.paritytech.polkadotapp.feature_products_impl.domain.truapi

import io.parity.truapi.PocketHostBridge
import io.paritytech.polkadotapp.feature_products_api.domain.pocket.PocketCard
import io.paritytech.polkadotapp.feature_products_api.domain.pocket.PocketCardId
import io.paritytech.polkadotapp.feature_products_api.domain.pocket.PocketCardKey
import io.paritytech.polkadotapp.feature_products_api.domain.pocket.PocketRemoval
import io.paritytech.polkadotapp.feature_products_api.domain.pocket.PocketRemoveError
import io.paritytech.polkadotapp.feature_products_api.model.ProductId
import io.paritytech.polkadotapp.feature_products_impl.domain.pocket.PocketCardStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import uniffi.truapi_server.NativePocketRemoval
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference
import uniffi.truapi.PocketCard as NativePocketCard

/**
 * Serves one product's slice of the collection to the core. Both callbacks run inline on the core's
 * dispatcher thread: the list is answered from a snapshot, and a removal completes before returning
 * because the core reads the list again right after it and republishes that answer.
 */
class ProductPocketHostBridge(
    private val productId: ProductId,
    private val store: PocketCardStore,
    private val scope: CoroutineScope,
) : PocketHostBridge {
    private val snapshot = AtomicReference<List<NativePocketCard>>(emptyList())
    private var collector: Job? = null
    private val stopped = AtomicBoolean(false)

    /**
     * Keeps the snapshot current and republishes the product's cards on every collection change.
     * Changes that leave this product's slice as it was are not republished: a face streaming at
     * frame rate changes the stored collection continuously without changing any card the core knows.
     */
    fun start(republish: (List<NativePocketCard>) -> Unit) {
        collector = scope.launch {
            store.observeCards()
                .map { cards -> cards.filter { it.key.productId == productId }.map { it.toNative() } }
                .distinctUntilChanged()
                .collect { cards ->
                    snapshot.set(cards)
                    // Cancelling does not wait, and the caller closes the execution as soon as it
                    // returns: a republish already in flight would otherwise reach a freed handle.
                    if (!stopped.get()) republish(cards)
                }
        }
    }

    /** Ends the republishing; [republish] is handed the execution, which its owner is about to close. */
    fun stop() {
        stopped.set(true)
        collector?.cancel()
        collector = null
    }

    override fun listCards(): List<NativePocketCard> = snapshot.get()

    override fun removeCard(cardId: String): NativePocketRemoval {
        // The snapshot carries the flag, so a card the host placed is refused without the blocking
        // read below, which runs on the core's own dispatcher thread.
        if (snapshot.get().any { it.cardId == cardId && it.privileged }) return NativePocketRemoval.PRIVILEGED

        val removal = runBlocking { store.removeCard(PocketCardKey(productId, PocketCardId(cardId))) }
        return removal.fold(
            onSuccess = { outcome ->
                snapshot.updateAndGet { cards -> cards.filterNot { it.cardId == cardId } }
                outcome.toNative()
            },
            onFailure = { failure ->
                if (failure is PocketRemoveError.Privileged) NativePocketRemoval.PRIVILEGED else throw failure
            },
        )
    }

    private fun PocketRemoval.toNative() = when (this) {
        PocketRemoval.REMOVED -> NativePocketRemoval.REMOVED
        PocketRemoval.ABSENT -> NativePocketRemoval.ABSENT
    }

    private fun PocketCard.toNative() = NativePocketCard(cardId = key.cardId.value, privileged = privileged)
}

File: hosts/ios/polkadot-app/Modules/Products/Pocket/ProductPocketHostBridge.swift
import Foundation
import os
import Products
import TrUAPIHost

/// Serves one product's slice of the collection to the core.
///
/// Both callbacks run inline on the core's dispatcher thread, so neither may
/// await: the list is answered from a snapshot, and a removal decides against
/// that same snapshot before touching storage.
final class ProductPocketHostBridge: PocketHostBridge, @unchecked Sendable {
    private let productId: String
    private let collection: any PocketCollection
    private let snapshot = OSAllocatedUnfairLock(initialState: [PocketCard]())
    private let republish = OSAllocatedUnfairLock(initialState: (([PocketCard]) -> Void)?.none)

    init(productId: String, collection: any PocketCollection) {
        self.productId = productId
        self.collection = collection
    }

    /// Starts republishing this product's cards to the core. `publish` is handed
    /// the execution, which its owner closes, so it is dropped on ``stop()``.
    func start(publish: @escaping ([PocketCard]) -> Void) {
        republish.withLock { $0 = publish }
    }

    func stop() {
        republish.withLock { $0 = nil }
    }

    /// Re-reads the collection and tells the core only if this product's own
    /// slice changed. A face streaming at frame rate changes the stored
    /// collection continuously without changing any card the core knows about.
    func refresh() async {
        let current = await collection.cards()
            .filter { $0.key.productId == productId }
            .map { PocketCard(cardId: $0.key.cardId.value, privileged: $0.privileged) }

        let changed = snapshot.withLock { held -> Bool in
            guard held != current else { return false }
            held = current
            return true
        }
        guard changed else { return }

        republish.withLock { $0 }?(current)
    }

    func listCards() throws -> [PocketCard] {
        snapshot.withLock { $0 }
    }

    func removeCard(cardId: String) throws -> NativePocketRemoval {
        // Answered from the snapshot: a card the host placed is refused without
        // the blocking read below, which runs on the core's own thread.
        if snapshot.withLock({ $0 }).contains(where: { $0.cardId == cardId && $0.privileged }) {
            return .privileged
        }

        let key = PocketCardKey(productId: productId, cardId: PocketCardId(value: cardId))
        let outcome = blockingRemove(key)

        if outcome == .removed {
            snapshot.withLock { $0.removeAll { $0.cardId == cardId } }
        }
        return outcome
    }

    /// The core waits on this answer, so the removal is completed here rather
    /// than handed to a task the caller cannot observe.
    private func blockingRemove(_ key: PocketCardKey) -> NativePocketRemoval {
        let result = OSAllocatedUnfairLock(initialState: NativePocketRemoval.absent)
        let done = DispatchSemaphore(value: 0)

        Task {
            let outcome: NativePocketRemoval
            do {
                outcome = try await collection.removeCard(key) == .removed ? .removed : .absent
            } catch PocketRemoveError.privileged {
                outcome = .privileged
            } catch {
                outcome = .absent
            }
            result.withLock { $0 = outcome }
            done.signal()
        }

        done.wait()
        return result.withLock { $0 }
    }
}
